### PR TITLE
Adjust asset job backfills to use asset backfill code path

### PR DESCRIPTION
**Pull request by smackesey**

### CODE REVIEW

**Comment by smackesey** [comment]
* **#21259** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21259?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @smackesey and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by schrockn** [comment]
I assume I should to wait to review until you figured out testing story and take out of drafts, correct @smackesey ?

File: python_modules/dagster/dagster/_core/definitions/job_definition.py [hunk 1]
```diff
@@ -416,6 +417,15 @@ def partitions_def(self) -> Optional[PartitionsDefinition]:
         """
         return None if not self.partitioned_config else self.partitioned_config.partitions_def
 
+    @property
+    def backfill_policy(self) -> Optional[BackfillPolicy]:
+        first_key = next(iter(self.asset_layer.executable_asset_keys), None)
+        if self.partitions_def is None or first_key is None:
+            return None
+        # It is enforced that all assets in a job have the same backfill policy, so we can just take
+        # the first backfill policy.
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/job_definition.py:426]: Consider an invariant to enforce this as a sanity check and using `@cached_property`

File: python_modules/dagster/dagster/_core/definitions/unresolved_asset_job_definition.py [hunk 2]
```diff
@@ -187,6 +190,32 @@ def resolve(
...
+        if len(backfill_policies) > 1:
+            raise DagsterInvalidDefinitionError(
+                f"Asset job {self.name} materializes asset with varying BackfillPolicies. All assets"
+                " in a job must share the same BackfillPolicy."
+            )
+
+        # Error if a PartitionedConfig is defined and any target asset has a backfill policy that
+        # materializes anything other than a single partition per run. This is because
+        # PartitionedConfig is a function that maps single partition keys to run config, so it's
+        # behavior is undefined for multiple-partition runs.
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/unresolved_asset_job_definition.py:206]: should we deprecate `PartitionedConfig`?

File: python_modules/dagster/dagster/_core/execution/job_backfill.py [hunk 3]
```diff
@@ -124,14 +131,32 @@ def _check_repo_has_partition_set(
...
+    instance: DagsterInstance,
+    max_range_size: Optional[int],
+):
+    if max_range_size is None:
+        return [partition_key_range]
+    else:
+        keys = partitions_def.get_partition_keys_in_range(
+            partition_key_range, dynamic_partitions_store=instance
+        )
+        chunks = [keys[i : i + max_range_size] for i in range(0, len(keys), max_range_size)]
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/execution/job_backfill.py:149]: this is comment-worthy

File: python_modules/dagster/dagster/_core/execution/job_backfill.py [hunk 4]
```diff
@@ -124,14 +131,32 @@ def _check_repo_has_partition_set(
...
+        return [PartitionKeyRange(start=chunk[0], end=chunk[-1]) for chunk in chunks]
 
 
 def _get_partitions_chunk(
     instance: DagsterInstance,
     logger: logging.Logger,
     backfill_job: PartitionBackfill,
     chunk_size: int,
-) -> Tuple[Sequence[str], str, bool]:
+    partition_set: ExternalPartitionSet,
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/execution/job_backfill.py:158]: would be a big fan of a rename to RemotePartitionSet while you are at it in this stack

File: python_modules/dagster/dagster/_core/execution/job_backfill.py [hunk 5]
```diff
@@ -151,28 +176,48 @@ def _get_partitions_chunk(
...
+
     return to_submit, next_checkpoint, has_more
 
 
 def submit_backfill_runs(
     instance: DagsterInstance,
     create_workspace: Callable[[], BaseWorkspaceRequestContext],
     backfill_job: PartitionBackfill,
-    partition_names: Optional[Sequence[str]] = None,
+    partition_names_or_ranges: Optional[Sequence[Union[str, PartitionKeyRange]]] = None,
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/execution/job_backfill.py:220]: should we change to partition keys for consistency?

File: python_modules/dagster/dagster/_core/execution/job_backfill.py [hunk 6]
```diff
@@ -212,7 +252,51 @@ def submit_backfill_runs(
...
+        external_repo.handle,
+        partition_set_name,
+        partition_data_target,
+        instance,
+    )
+    assert isinstance(partition_set_execution_data, ExternalPartitionSetExecutionParamData)
+
+    # Partition-scoped run config is prohibited at the definitions level for a jobs that materialize
+    # ranges, so we can assume that all partition data will have the same run config and tags as the
+    # first partition.
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/execution/job_backfill.py:272]: an invariant to actually verify this assumption would be very reassuring here.

File: python_modules/dagster/dagster/_core/execution/job_backfill.py [hunk 7]
```diff
@@ -212,7 +252,51 @@ def submit_backfill_runs(
...
+    tags_by_key_or_range: Mapping[Union[str, PartitionKeyRange], Mapping[str, str]]
+    run_config_by_key_or_range: Mapping[Union[str, PartitionKeyRange], Mapping[str, Any]]
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/execution/job_backfill.py:274]: At this point coming up with a name for `Union[str, PartitionKeyRange]` and using it in this code would be very helpful.

Alternatively you could canonicalize to `PartitionKeyRange`, which is probably preferable.

Comment (schrockn) [reply python_modules/dagster/dagster/_core/execution/job_backfill.py:274]: You could have to have a class that "remembers" its lineage to create the correct tags (and maybe we should consider canonicalizing that as well?).

**Review by schrockn — CHANGES_REQUESTED** [review]
This is a good start. 

I do think this PR should include a detailed test plan that documents the new behavior with screenshots etc to demonstrate the before/after in terms of user experience.

File: python_modules/dagster/dagster/_core/definitions/unresolved_asset_job_definition.py (hunk 2 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/definitions/unresolved_asset_job_definition.py:206]: The answer to this depends on the broader question of the future of job backfills and partitioned jobs.

File: python_modules/dagster/dagster/_core/execution/job_backfill.py (hunk 4 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/execution/job_backfill.py:158]: Have got that in my queue. Big stack of renames I will land this week.

File: python_modules/dagster/dagster/_core/execution/job_backfill.py (hunk 7 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/execution/job_backfill.py:274]: >Alternatively you could canonicalize to PartitionKeyRange, which is probably preferable.

Yes, my plan was to normalize the entire job backfill pathway to `PartitionKeyRange` in a follow-up as a starting point for this workstream, since it is a relatively "contained" body of code that I now understand very well.

File: python_modules/dagster/dagster/_core/execution/job_backfill.py (hunk 6 above)
Comment (schrockn) [reply python_modules/dagster/dagster/_core/execution/job_backfill.py:276]: Got it thanks

**Review by schrockn — APPROVED** [review]
Lovely

**Comment by smackesey** [comment]
Hi @the4thamigo-uk, thanks for the report, we're aware of this issue. What's happening is that `dagster-dbt` actually sets `BackfillPolicy.single_run()` as the default if `None` is passed, which is leading to this problem. We're looking for a better solution but in the meantime you'll need to make sure all the partitioned assets in the job have the same backfill policy. You can do this either by setting your other partitioned assets in the job to use `BackfillPolicy.single_run()` or setting all assets in the job to use some other policy.

**Comment by the4thamigo-uk** [comment]
Thanks @smackesey good to know you are already on the case. I have already worked around this by setting the assets to single_run() like you suggest.

**Comment by the4thamigo-uk** [comment]
> Hi @the4thamigo-uk, thanks for the report, we're aware of this issue. What's happening is that `dagster-dbt` actually sets `BackfillPolicy.single_run()` as the default if `None` is passed, which is leading to this problem. We're looking for a better solution but in the meantime you'll need to make sure all the partitioned assets in the job have the same backfill policy. You can do this either by setting your other partitioned assets in the job to use `BackfillPolicy.single_run()` or setting all assets in the job to use some other policy.

@smackesey This is fine apart from the fact that changing the backfill policy of my non-dbt partitioned assets is not straight forward, as they need to support processing of multiple partitions according to the comments in https://docs.dagster.io/concepts/partitions-schedules-sensors/backfills#launching-single-run-backfills-using-backfill-policies-

**Comment by the4thamigo-uk** [comment]
Hi @smackesey, Im not explicitly assigning any backfill policies for any assets, and none of them support `single_run`. The issue I'm facing is that dagster-dbt sets the policy to `single_run` and this conflicts with my other assets which have no policy set.

**Comment by smackesey** [comment]
>Hi @smackesey, Im not explicitly assigning any backfill policies for any assets, and none of the non-dbt assets support single_run. The issue I'm facing is that dagster-dbt sets the policy to single_run and this conflicts with my other assets which have no policy set.

Right-- let me rephrase my question. Do your DBT assets actually support single-run, or do they just have that set as their backfill policy? Since it's the default that you didn't set, and you've been executing these assets as part of a job (which ignored backfill policy until the fix a few weeks ago), I can see a world where your DBT assets don't even _actually_ support single-run, regardless of what the backfill policy says.

**Comment by the4thamigo-uk** [comment]
Yeah I dont think it ever supported it...

**Comment by the4thamigo-uk** [comment]
> Yeah I dont think it ever supported it...

I think what I said might not be strictly correct, I am using the `context.partition_time_window` to obtain the date range to feed to the `dbt build` command, and this should support `single_run` according to https://docs.dagster.io/concepts/partitions-schedules-sensors/backfills#launching-single-run-backfills-using-backfill-policies-.

However, I never used this for more than a single partition, as my partitions were always executed in separate job runs.

**Comment by the4thamigo-uk** [comment]
@smackesey by the way the latest release fixed this issue for me.
